staging/src/k8s.io/component-helpers/resource/helpers.go

```diff
@@ -0,0 +1,268 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package resource
+
+import (
+	v1 "k8s.io/api/core/v1"
+)
+
+// PodResourcesOptions controls the behavior of PodRequests and PodLimits.
+type PodResourcesOptions struct {
+	// Reuse, if provided will be reused to accumulate resources and returned by the PodRequests or PodLimits
+	// functions. All existing values in Reuse will be lost.
+	Reuse v1.ResourceList
+	// InPlacePodVerticalScalingEnabled indicates that the in-place pod vertical scaling feature gate is enabled.
+	InPlacePodVerticalScalingEnabled bool
+	// ExcludeOverhead controls if pod overhead is excluded from the calculation.
+	ExcludeOverhead bool
+	// ContainerFn is called with the effective resources required for each container within the pod.
+	ContainerFn func(res v1.ResourceList, containerType ContainerType)
+	// NonMissingContainerRequests if provided will replace any missing container level requests for the specified resources
+	// with the given values.  If the requests for those resources are explicitly set, even if zero, they will not be modified.
+	NonMissingContainerRequests v1.ResourceList
+}
+
+// PodRequests computes the pod requests per the PodResourcesOptions supplied. If PodResourcesOptions is nil, then
+// the requests are returned including pod overhead. The computation is part of the API and must be reviewed
+// as an API change.
+func PodRequests(pod *v1.Pod, opts PodResourcesOptions) v1.ResourceList {
+	// attempt to reuse the maps if passed, or allocate otherwise
+	reqs := reuseOrClearResourceList(opts.Reuse)
+
+	var containerStatuses map[string]*v1.ContainerStatus
+	if opts.InPlacePodVerticalScalingEnabled {
+		containerStatuses = make(map[string]*v1.ContainerStatus, len(pod.Status.ContainerStatuses))
+		for i := range pod.Status.ContainerStatuses {
+			containerStatuses[pod.Status.ContainerStatuses[i].Name] = &pod.Status.ContainerStatuses[i]
+		}
+	}
+
+	for _, container := range pod.Spec.Containers {
+		containerReqs := container.Resources.Requests
```

Comment:
Question for @liggitt: would it be fine to also incorporate the logic of "requests = limits, when not defined".

https://github.com/kubernetes/kubernetes/blob/b041e54e70fc906eb753b6acc6967530cd3fb720/pkg/apis/core/v1/defaults.go#L158

Reply:
two thoughts:
1. don't mix relocation and modification into the same PR
2. why would we need to duplicate API defaulting logic here? admission / controllers / clients reading from the API will already see pods with the defaults applied

Reply:
1. I agree.
2. Because users of this helper might be using a Pod template that didn't pass through the apiserver yet. For example, when using an external CRD such as a kubeflow TFJob.

Reply:
but... the input to this function is Pod, not PodTemplate

Reply:
Yeah, but one can easily convert one to the other in order to get a resource calculation.

Reply:
If anything, let's leave this for a follow up.

Reply:
Another follow up question: Is there a reason whey PodRequests has the entire Pod object as the parameter instead of PodSpec and PodStatus explicitly? I wanted to modify the Validation logic of PodTemplate to use PodRequests to validate pod-level resources validation criteria and hence wondering...